File: use-cases/school-timetabling/src/test/java/org/acme/schooltimetabling/rest/TimeTableResourceTest.java
package org.acme.schooltimetabling.rest;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import javax.inject.Inject;

import org.acme.schooltimetabling.domain.Lesson;
import org.acme.schooltimetabling.domain.TimeTable;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.optaplanner.core.api.solver.SolverStatus;

import io.quarkus.test.junit.QuarkusTest;

@QuarkusTest
public class TimeTableResourceTest {

    @Inject
    TimeTableResource timeTableResource;

    @Test
    @Timeout(600_000)
    public void solveDemoDataUntilFeasible() throws InterruptedException {
        timeTableResource.solve();
        TimeTable timeTable;
        do { // Use do-while to give the solver some time and avoid retrieving an early infeasible solution.
            // Quick polling (not a Test Thread Sleep anti-pattern)
            // Test is still fast on fast machines and doesn't randomly fail on slow machines.
            Thread.sleep(20L);
            timeTable = timeTableResource.getTimeTable();
        } while (timeTable.getSolverStatus() != SolverStatus.NOT_SOLVING || !timeTable.getScore().isFeasible());

        assertFalse(timeTable.getLessonList().isEmpty());
        for (Lesson lesson : timeTable.getLessonList()) {
            assertNotNull(lesson.getTimeslot());
            assertNotNull(lesson.getRoom());
        }
        assertTrue(timeTable.getScore().isFeasible());
    }

}

File: kotlin-quarkus-school-timetabling/src/test/kotlin/org/acme/kotlin/schooltimetabling/rest/TimeTableResourceTest.kt
/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.acme.kotlin.schooltimetabling.rest

import io.quarkus.test.junit.QuarkusTest
import org.acme.kotlin.schooltimetabling.domain.TimeTable
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Timeout
import org.optaplanner.core.api.solver.SolverStatus
import javax.inject.Inject


@QuarkusTest
class TimeTableResourceTest {

    @Inject
    lateinit var timeTableResource: TimeTableResource

    @Test
    @Timeout(600000)
    @Throws(InterruptedException::class)
    fun solveDemoDataUntilFeasible() {
        timeTableResource.solve()
        var timeTable: TimeTable = timeTableResource.getTimeTable()
        while (timeTable.solverStatus != SolverStatus.NOT_SOLVING) {
            // Quick polling (not a Test Thread Sleep anti-pattern)
            // Test is still fast on fast machines and doesn't randomly fail on slow machines.
            Thread.sleep(20L)
            timeTable = timeTableResource.getTimeTable()
        }
        Assertions.assertFalse(timeTable.lessonList.isEmpty())
        for (lesson in timeTable.lessonList) {
            Assertions.assertNotNull(lesson.timeslot)
            Assertions.assertNotNull(lesson.room)
        }
        Assertions.assertTrue(timeTable.score!!.isFeasible)
    }
    
}
